cEP-0028.md

```diff
@@ -0,0 +1,150 @@
+
+## Proposal
+
+The optimization process focuses to maintain minimal changes to coala core
+and rather optimize the coala and language server interface. The following
+are the proposed ideas.
+
+1. Extending coala configuration parameters
+   Some bears are not often required to be executed on every cycle of coala
+   analysis especially during use cases such as this. The parameter set can
+   thus be extended to support parameters of type `cls_*`. Since these
+   parameters are not associated with any particular bear but rather control
+   the effect of various sections during executing coala, no support extension
+   for bears or coala-core would be necessary. These parameters enable support
+   for actions such as enabling/disabling a section for coala language server
+   or hooking a section for a particular request type. For example:
+
+   ```
+    [longrunning]
+    enabled = True
+    cls_enabled = False
+    bears = SomeLongRunningBear
+   ```
+
+   This kind of a configuration will disable the bear completely during
+   analysis cycles of coala via a language server while enabling it for
+   regular invocations.
+
+   ```
+    [longrunning]
+    enabled = False
+    cls_on = save
+    files = *.py
+    bears = SomeLongRunningBear
+   ```
+
+   This example shows a configuration where the section is disabled for coala
+   but will only be enabled when the code editor dispatches a request of type
+   `textDocuemnt/didSave`.
+
+   ### Supported Parameters
+
+   The following are the supported parameter types:
+
+   - `cls_enabled`: A boolean indicating if the section should be enabled.
+     Disabling the section will turn if off for coala lang server. It will
+     default to true.
+
+   - `cls_on`: A list containing some or all of the available options,
+     `every`, `open`, `save`, `change`, `formatting`. This indicates the
+     type of request(s) the section should be executed on. `cls_on` will
+     default to `all`.
+
+   - `cls_fixes_enabled`: A boolean value indicating if the code fixes
+     generated by the section should be sent to the editor for being applied.
+     Initially only `formatting` request will respond with code fixes.
+
+   - `cls_as_global`: A boolean flag indicating if the section should be
+     considered as a global section. A global section executes analysis on
+     all of the files in its `files` attribute and reports diagnostic messages
+     on all the corresponding files open in the editor. If `cls_as_global` is
+     `false` then analysis runs only on the file currently being edited.
+     `cls_as_global` will default to false.
+
+2. Custom entry point into coala core
+   Since coala core was not designed to be a long running process coalib
+   does not provide a flexible entry point into it. To maximize the efficiency
+   of the coala language server, it will require an entry point that supports:
+
+   - Support for `FileProxies`: Since not all changes on the editor are flushed
```

Comment:
If these `FileProxies` are related to the core we are calling them `FileFactory` for now. It has been merged recently https://github.com/palash25/coala/blob/master/coalib/io/FileFactory.py

Reply:
@palash25 FileProxies are not FileFactories. FileProxies are being used to handle unflushed changes on files in coala language server. Although on a higher level they serve similar purposes,  FileProxies are custom built for coala-ls. Refer to https://github.com/ksdme/coala-ls/blob/gsoc-phase-1/coalals/utils/files.py#L59 for some insight.